File: src/Board.cpp
#include "Board.h"
#include <cmath>

Board::Board(int size) : size(size), totalNodes(0) {
    grid.resize(size, std::vector<Tile>(size));
}

Tile& Board::GetTile(int x, int y) { return grid[y][x]; }
const Tile& Board::GetTile(int x, int y) const { return grid[y][x]; }

Tile& Board::GetTile(GridPos pos) { return grid[pos.y][pos.x]; }
const Tile& Board::GetTile(GridPos pos) const { return grid[pos.y][pos.x]; }

bool Board::IsValidPos(GridPos pos) const {
    return pos.x >= 0 && pos.x < size && pos.y >= 0 && pos.y < size;
}

bool Board::IsAdjacent(GridPos a, GridPos b) const {
    return (std::abs(a.x - b.x) == 1 && a.y == b.y) || 
           (std::abs(a.y - b.y) == 1 && a.x == b.x);
}

std::vector<GridPos> Board::GetAdjacent(GridPos pos) const {
    std::vector<GridPos> adj;
    GridPos dirs[] = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    for (auto d : dirs) {
        GridPos p = {pos.x + d.x, pos.y + d.y};
        if (IsValidPos(p)) {
            adj.push_back(p);
        }
    }
    return adj;
}

void Board::ClearPaths() {
    for (int y = 0; y < size; y++) {
        for (int x = 0; x < size; x++) {
            if (grid[y][x].type == TileType::PATH) {
                grid[y][x].type = TileType::EMPTY;
                grid[y][x].sequenceNum = 0;
                grid[y][x].color = BLANK;
            }
        }
    }
}

void Board::ResetErrors() {
    for (int y = 0; y < size; y++) {
        for (int x = 0; x < size; x++) {
            grid[y][x].isError = false;
        }
    }
}

bool Board::IsFull() const {
    for (int y = 0; y < size; y++) {
        for (int x = 0; x < size; x++) {
            if (grid[y][x].type == TileType::EMPTY) return false;
        }
    }
    return true;
}

void Board::ResetBoard() {
    ClearPaths();
    ResetErrors();
}


void Board::SetNode(GridPos pos, int sequenceNum, Color color) {
    if (IsValidPos(pos)) {
        grid[pos.y][pos.x] = {TileType::NODE, sequenceNum, color, false};
    }
}

void Board::SetPath(GridPos pos, int sequenceNum, Color color) {
    if (IsValidPos(pos)) {
        grid[pos.y][pos.x] = {TileType::PATH, sequenceNum, color, false};
    }
}

void Board::ClearTile(GridPos pos) {
    if (IsValidPos(pos)) {
        grid[pos.y][pos.x] = {TileType::EMPTY, 0, BLANK, false};
    }
}

File: lib/models/board.dart
import 'package:flutter/material.dart';
import 'grid_pos.dart';
import 'tile.dart';

class Board {
  final int size;
  int totalNodes;
  late List<List<Tile>> grid;

  // Stores walls between adjacent cells. Key format: "minX,minY-maxX,maxY"
  final Set<String> walls = {};

  Board(this.size) : totalNodes = 0 {
    grid = List.generate(size, (_) => List.generate(size, (_) => Tile()));
  }

  String _wallKey(GridPos a, GridPos b) {
    if (a.x < b.x || (a.x == b.x && a.y < b.y)) {
      return '${a.x},${a.y}-${b.x},${b.y}';
    } else {
      return '${b.x},${b.y}-${a.x},${a.y}';
    }
  }

  void addWall(GridPos a, GridPos b) {
    if (isValidPos(a) && isValidPos(b)) {
      walls.add(_wallKey(a, b));
    }
  }

  bool hasWall(GridPos a, GridPos b) {
    return walls.contains(_wallKey(a, b));
  }

  Tile getTile(GridPos pos) {
    if (!isValidPos(pos)) return Tile();
    return grid[pos.y][pos.x];
  }

  bool isValidPos(GridPos pos) {
    return pos.x >= 0 && pos.x < size && pos.y >= 0 && pos.y < size;
  }

  /// Checks if two cells are geometrically adjacent AND not separated by a wall.
  bool isAdjacent(GridPos a, GridPos b) {
    bool geomAdj = (a.x == b.x && (a.y - b.y).abs() == 1) ||
        (a.y == b.y && (a.x - b.x).abs() == 1);
    if (!geomAdj) return false;
    return !hasWall(a, b);
  }

  List<GridPos> getAdjacent(GridPos pos) {
    List<GridPos> adj = [];
    final offsets = [
      GridPos(0, -1),
      GridPos(0, 1),
      GridPos(-1, 0),
      GridPos(1, 0),
    ];
    for (var offset in offsets) {
      final newPos = GridPos(pos.x + offset.x, pos.y + offset.y);
      if (isValidPos(newPos)) {
        adj.add(newPos);
      }
    }
    return adj;
  }

  void clearTile(GridPos pos) {
    if (isValidPos(pos)) {
      grid[pos.y][pos.x] = Tile();
    }
  }

  void setNode(GridPos pos, int sequenceNum, Color color) {
    if (isValidPos(pos)) {
      grid[pos.y][pos.x] = Tile(
        type: TileType.node,
        sequenceNum: sequenceNum,
        color: color,
      );
    }
  }
}
